File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return help(preorder,inorder,0,inorder.length-1,new int[]{0});
    }
    public TreeNode help(int[] preorder,int[] inorder,int l,int r,int[] ind){
        if(l>r)return null;
        TreeNode root=new TreeNode(preorder[ind[0]]);
        int index=find(inorder,preorder[ind[0]]);
        ind[0]++;
        root.left=help(preorder,inorder,l,index-1,ind);
        root.right=help(preorder,inorder,index+1,r,ind);
        return root;
    }
    public int find(int[] arr,int val){
        for(int i=0;i<arr.length;i++){
            if(arr[i]==val)return i;
        }
        return -1;
    }
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i=0;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder,inorder,0,preorder.size()-1);
    }
    TreeNode* build(vector<int>& preorder, vector<int>& inorder,int l,int r){
        if(l>r)return NULL;
        TreeNode* newNode=new TreeNode(preorder[i]);
        int ind=search(inorder,preorder[i++]);
        newNode->left=build(preorder,inorder,l,ind-1);
        newNode->right=build(preorder,inorder,ind+1,r);
        return newNode;
    }
    int search(vector<int>& arr,int n){
        for(int i=0;i<arr.size();i++){
            if(arr[i]==n)return i;
        }
        return -1;
    }
};